Fix multi-shard SELECT to pull only needed columns

jasonmp85 opened the pull request:
Reworks multi-shard SELECT so instead of performing quals both remotely and on the local sequential scan, they are only performed remotely. This frees us up to avoid pulling columns mentioned only in quals over the wire, which provides users more control over how much data gets pulled in a multi-shard query.

fixes #33

sumedhpathak commented on pg_shard.c:
```diff
@@ -625,8 +634,13 @@ PlanSequentialScan(Query *query, int cursorOptions, ParamListInfo boundParams)
+	 * apply quals for us: RowAndColumnFilterQuery ensures all quals are pushed
+	 * down during remote executions. The mutator returns a copy of the query.
+	 */
+	modifiedQuery = (Query *) RemoveFromExpressionQualsMutator((Node *) query, NULL);
```
Why do we need the mutator construct here? Isn't setting the quals to NIL not sufficient?

jasonmp85 replied:
The `FromExpr` is actually a tree, so recursing down it to set all `quals` to `NULL` is most appropriate. I believe that we shouldn't ever see anything other than a lone node as we're not supporting `JOINs` at the moment, but it seemed more "correct" to recurse along the entire tree.

I could set `quals` to `NULL` and just document with a `TODO` or something that we would need to change this piece of code if we ever push down `JOIN`s.…

sumedhpathak commented on pg_shard.c:
```diff
@@ -813,6 +818,46 @@ ExtractFromExpressionWalker(Node *node, List **qualifierList)
+ * from any FromExpr nodes it encounters. Returns a copy of the original query.
+ */
+static Node *
+RemoveFromExpressionQualsMutator(Node *originalNode, void *context)
```
This PR then doesn't handle the mutable clauses use-case? It appears to remove all the clauses from local evaluation?

jasonmp85 replied:
Correct. At the moment I'm removing _all_ local clauses and continue to send all clauses for remote execution.

#47 uses the word `VOLATILE` in its description, but the intent is closer to what you just described. Because the fix for what you describe needs to address the problem for `UPDATE` and `DELETE` statements as well (and possibly `INSERT`?) it will be better handled in a PR for #47 which will be focused entirely on that problem.

sumedhpathak commented:
We discussed about modifying the code to make it closer to postgres_fdw, which means we break up quals into 'local' and 'remote'.
The remote query would then pull columns for the local quals, and execute the remote quals entirely. This makes the logic cleaner, and sets it up for handling volatile/mutable functions. To begin with, we expect the local quals to be NIL as we push all of them down currently.

We also discussed using a mutator for modifying the FromExpr for the local quals, vs setting the quals to NULL. We decided on taking the simpler approach now, acknowledging that this may need to be revisited with subselect handling.

jasonmp85 commented:
I haven’t written the `ClassifyRestrictions` function yet, though at the moment it will trivially put all quals in the remote list and leave the local list empty. So all the functionality is there to use these variables (even if the local quals eventually have something to do), but that function is absent (as is documentation).

I didn’t want to change PlanSequentialScan’s parameter list as they’re analogous to the ordering of params in other `plan*` functions in PostgreSQL.

Finally, I want to point out that our old code had weird requirements around what received the original `query` and what received `distributedQuery`, and in fact certain lists and objects build using `distributedQuery` were later used with `query` to build the sequential scan, which felt weird. I retained all those requirements, as they're actually necessary, but it bears calling out that `distributedQuery` has had mutations performed on it by `standard_planner` which means the final `localQuery` is a bit of an amalgam of computed and raw things. Again, this is preserving existing behavior of `pg_shard` but I wanted to mention it.

sumedhpathak commented on pg_shard.c:
```diff
@@ -937,6 +973,12 @@ BuildDistributedPlan(Query *query, List *shardIntervalList)
+		if (LogDistributedStatements)
+		{
+			ereport(LOG, (errmsg("distributed statement: %s", queryString->data),
+						  errhidestmt(true)));
```
The comment for errhidestmt mentions that it can be used for error messages which contain the statement. This one doesn't contain the original statement (it contains the transformed query that we send down to the worker nodes). Is it useful to retain that, so we can tie which transformed statement was a result of which original user query?

jasonmp85 replied:
I _think_ this is so disparate callers can ensure a given _STATEMENT:_ line is only emitted once per stack level. I've looked into my use of it here and it serves no purpose. I'm not suppressing the _STATEMENT:_ line, nor am I trying to print more than one line per shard, so it's not suppressing anything ever. I'll remove it.

jasonmp85 replied:
BTW the config to emit the "main" query is `log_statement` and is orthogonal to this line.

sumedhpathak commented on pg_shard.c:
```diff
@@ -641,14 +679,13 @@ PlanSequentialScan(Query *query, int cursorOptions, ParamListInfo boundParams)
-	List *whereClauseColumnList = NIL;
-	List *projectColumnList = NIL;
 	List *columnList = NIL;
+	List *localColumnList = NIL;
```
Can we think of other, more descriptive names here? LocalColumnList doesn't immediately indicate to me what it means. What about retaining whereClauseColumnList?

jasonmp85 replied:
They were named after `localConds` and `remoteConds` in `postgres_fdw`. I find the `local`/`remote` distinction _very_ explicit/clear, as it immediately tells me what a variable is used for (local or remote execution). `whereClause` says where we got the value, not what it is for.

But while writing this I remember wondering if the naming couldn't be better, so I want to dump all the `local`/`remote` names here and we can discuss which need changing:

  * `localQuery` — The `Query` structure used to build a local sequential scan
  * `localRestrictList` — A list of expressions from `WHERE` clauses that must be included in the local sequential scan. Currently always `NIL`.
  * `localColumnList` — The full list of columns referenced by the `localRestrictList` quals
  * `remoteRestrictList` — A list of expressions from `WHERE` clauses that will be executed remotely. Currently contains all `WHERE` clauses from the original query.

Was it just `localColumnList` you had a problem with, or are any of the others problematic? Some of these occur both as variable names and as parameter names.

sumedhpathak replied:
I think `localQuery`, `localRestrictList` and `remoteRestrictList` all make sense. I think the attribute of `local` to a column isn't immediately descriptive to me. I guess the long form would be localRestrictClauseColumnList? 

In the context of this function, I was thinking the two sources of columns to be fetched: projectColumns and whereClauseColumns. This duality to me makes it a bit clearer. The combination of these two lists from the toBeFetchedColumnList.

What do you think?

jasonmp85 replied:
`toBeFetchedColumnList` is a bit unwieldy, but the others seem great. Going to do that.

jasonmp85 replied:
Calling it `requiredColumnList`, as they're required to service the query.

sumedhpathak commented on pg_shard.c:
```diff
@@ -703,8 +739,8 @@ RowAndColumnFilterQuery(Query *query)
 	filterQuery->rtable = rangeTableList;
-	filterQuery->targetList = targetList;
 	filterQuery->jointree = fromExpr;
+	filterQuery->targetList = targetList;
```
Any reason for this change? (Changing the ordering)

jasonmp85 replied:
Yes. The above blocks build `rangeTableList` then `fromExpr` then `targetList`. I felt the assignments should mirror the order of construction.

sumedhpathak commented on pg_shard.c:
```diff
@@ -92,9 +95,11 @@ static Oid ExtractFirstDistributedTableId(Query *query);
 static bool ExtractRangeTableEntryWalker(Node *node, List **rangeTableList);
 static List * DistributedQueryShardList(Query *query);
 static bool SelectFromMultipleShards(Query *query, List *queryShardList);
+static Query * BuildLocalQuery(Query *query, List *localRestrictList);
```
I've used a heuristic of trying to maintain the order based on order of the call to the function. So in this case RowAndColumnFilterQuery is being called before BuildLocalQuery. Could you maintain that heuristic (and also make sure the forward declaration order is the same as the implementation order?

jasonmp85 replied:
I indeed wanted to change the order but knew doing so would dirty the diff (it would just highlight the whole method in green). You could always view each commit separately, but that makes commenting harder. I'm going to clean them up in exactly this way once we hash out the implementation issues first.

sumedhpathak commented on pg_shard.c:
```diff
@@ -179,6 +184,11 @@ _PG_init(void)
 							 &UseCitusDBSelectLogic, false, PGC_USERSET, 0, NULL,
 							 NULL, NULL);
 
+	DefineCustomBoolVariable("pg_shard.log_distributed_statements",
```
Naming: log_remote_query? log_shard_query? What do you think of those options?

jasonmp85 replied:
The logic I used to get me `log_distributed_statements`

  * Some settings begin with `debug_` and others with `log_`, though the debug ones tend to be about logging a complicated structure, not one line. So I went with `log_`
  * The (probably well-known) configuration for PostgreSQL to log every query is `log_statement`. I therefore wanted to use _statement_ in my name
  * As an end-user I understand _query_ to mean a `SELECT`: something that asks for data and returns it. A _statement_, however, has side-effects
  * `distributed` is because these statements are fields in the distributed plan. I'm logging them at the point of planning, not execution, so they are not yet remote (`log_statement` also logs far before execution)
  * Finally, it's plural because unlike `log_statement`, we're going to have multiple statements per plan

So in short, I don't like _query_ (inappropriate for `INSERT`, `UPDATE`, or `DELETE`) and `remote` makes sense if I'm logging in the course of remote execution, which I'm not. Tell me if this isn't convincing.

sumedhpathak replied:
Hm, I see. I was under the impression this was only being logged for SELECT's, but looks like its logging even updates and deletes. I'm convinced then to leave it as it is.

sumedhpathak commented on pg_shard.c:
```diff
@@ -246,15 +256,26 @@ PgShardPlanner(Query *query, int cursorOptions, ParamListInfo boundParams)
+			List *remoteRestrictList = NIL;
+			List *localRestrictList = NIL;
+
+			/* TODO: replace with call to classify restrictions */
```
Clarification: Do you intend to make this change in the current PR? Or is this for the future?

jasonmp85 replied:
Right now.

sumedhpathak commented on pg_shard.c:
```diff
@@ -246,15 +256,26 @@ PgShardPlanner(Query *query, int cursorOptions, ParamListInfo boundParams)
+			remoteRestrictList = queryRestrictList;
+
+			/* build local and distributed query */
+			distributedQuery = RowAndColumnFilterQuery(distributedQuery,
```
Should we rename to remoteQuery as we're using localQuery? local/remote have a more natural pairing than local/distributed.

jasonmp85 replied:
Hm. I honestly don't know. The problem is the object is a `DistributedPlan`, so there's a duality in that we build the `distributedPlan` using the `distributedQuery` but use the `localQuery` to build a local plan (the sequential scan, currently just named `plannedStatement`). I'll have to think a bit more about this.

jasonmp85 replied:
Chatted and keeping as `distributedQuery`.

sumedhpathak commented on pg_shard.c:
```diff
@@ -586,6 +607,25 @@ SelectFromMultipleShards(Query *query, List *queryShardList)
+ * TODO: Add documentation
+ */
+static Query *
+BuildLocalQuery(Query *query, List *localRestrictList)
```
Function naming: Do we expect this function to do anything apart from updating the where-clause list? If so, should we use a more specific name here? Something like UpdateRestrictClauses()?

jasonmp85 replied:
It doesn't modify the input, it returns a modified copy (this ties into the question about `copyObject` later).

sumedhpathak commented on pg_shard.c:
```diff
@@ -586,6 +607,25 @@ SelectFromMultipleShards(Query *query, List *queryShardList)
+	Query *localQuery = copyObject(query);
+	FromExpr *joinTree = localQuery->jointree;
+
+	if (joinTree != NULL)
```
Should this be an assert? If the joinTree is NULL, we aren't updating the clauses, but returning the query as-is. We should either: a) Error out if we expect joinTree to be NULL, b) Assert if we expect it to be there, c) Create a joinTree if its NULL and set the clauses?

jasonmp85 replied:
You're right: I was focusing on the `fromlist` having exactly one entry, but we can also expect `joinTree` to never be null. Going to change to assert (don't want to error as this should not be a user-facing problem: `Assert` is more appropriate).

sumedhpathak commented on pg_shard.c:
```diff
@@ -625,8 +664,7 @@ PlanSequentialScan(Query *query, int cursorOptions, ParamListInfo boundParams)
 
-	queryCopy = copyObject(query);
-	sequentialScanPlan = standard_planner(queryCopy, cursorOptions, boundParams);
+	sequentialScanPlan = standard_planner(query, cursorOptions, boundParams);
```
So the reason we kept this here is because the standard_planner scribbles on the query input. This way, I know PlanSequentialScan will keep my inputs unchanged. Removing this means the caller needs to be aware of this, and create a copy of the query in advance before calling this function. I'd be in favor of keeping it here.

jasonmp85 replied:
`BuildLocalQuery` creates a copy for us, but I didn't want to put its logic inside of `PlanSequentialScan`. So we're left with either:

  * Having two calls to `copyObject` even though only one is needed
  * Moving `BuildLocalQuery`'s logic inside of `PlanSequentialScan`, which complicates the signature and single responsibility principle.

What if I just change the function comment on `PlanSequentialScan` to document that it expects to "own" `query` and will modify it?

sumedhpathak replied:
Hmm, so now since we modify the localQuery, it makes sense to make a copy of it before we update the quals to NIL. Lets change the function comment here to document that it modifies the inputs. On looking at it a bit more, I like that the planner modifications to the localQuery which generated the plannedStmt are available to the caller of this function.

jasonmp85 replied:
Actually updated the comment already before you wrote this reply, so I think this is covered now.

jasonmp85 commented:
Responded to these pieces of feedback:
  * Remove use of `errhidestmt`
  * Added Assert that `query->jointree` is never `NULL`
  * Documented `BuildLocalQuery`
  * Clarified that `PlanSequentialScan` modifies its `query` argument
  * Added `ClassifyRestrictions` function

Needing further feedback:

  - [ ] Clarification on `local`/`remote` naming, as it's somewhat pervasive
  - [ ] How against `log_distributed_statements` are you? I like it and find alternatives less clear
  - [ ] How should I address the `copyObject` question? I added comments clarifying that one function _is_ returning a copy and another will scribble on input, but I don't see any other solution without adding an extraneous copy or adding too much responsibility to `PlanSequentialScan`. What was your suggestion?
  - [ ] Look over `ClassifyRestrictions`. It's naïve for now but I'll fill it in as part of #47
  - [ ] Is _restriction_ the right word? I like _qual_ but it feels informal (and _qualifier_ or _qualification_ feel too formal). What do we call the expressions in a `WHERE` clause? Couldn't find clear precedent in Citus…

jasonmp85 commented:
(And as mentioned, I'll modify function declaration/definition order as a last step).

sumedhpathak commented:
I think `restriction` is OK for now. We already have a function called QueryRestrictList, and it has precedence in Postgres. I think we've used whereClause or selectClause in Citus previously.
